_posts/2023-07-10-introducing-extensions-for-opensearch.md

```diff
@@ -0,0 +1,127 @@
+
+```json
+{
+    ...
+    "hits": {
+        "total": {
+            "value": 1000000,
+            "relation": "eq"
+        },
+        ...
+    ...
+}
+```
+
+The plugin limits these analysis rates to ensure stable cluster performance. After some fine-tuning (we increased simultaneous detector tasks from 10 to 96 and reduced the batch task piece interval from 5 to 3 seconds), the extension version of AD was able to successfully complete this analysis in 55 seconds. This isn’t quite apples-to-apples because the test in the previously mentioned blog post involved 1 million smaller models x 1 time slice each to generate 1 million results (we needed more time to implement support for real-time AD in the extension). However, we believe that we should be able to support 1M models on a single node or on a small number of nodes once extensions are capable of running in that environment.
+
+Because our extension ran on a remote node, we did observe an expected increase in API call latency. For example, the start/stop detector call went from 50--100ms to 200--300ms. These calls are marginal CRUD operations that are used to set up anomaly detection. The following figure provides a comparison in latency between extensions and plugins. For more information, see [Performance testing for the AD extension](https://github.com/opensearch-project/opensearch-sdk-java/issues/725#issuecomment-1597874382).
+
+![Extensions compared to plugins latency graph](/assets/media/blog-images/2023-07-10-introducing-extensions-for-opensearch/extension-graph.png){:.img-fluid}
+
+## Improving performance
+
+We proved that upgrading AD from plugin to extension could match existing performance for end users. But could we do better? We can, in the following three ways.  
+
+### 1. Use a cheaper cluster
+
+We found that by rewriting AD as an extension we could reduce the cost of our cluster and still meet the same performance bar. Because AD no longer runs on data nodes, we were able to replace all memory-optimized r5.2xlarge nodes with general purpose c5.2xlarge nodes for a 33% cost saving per data node for historical analysis, with the additional cost of one extension node.
+
+We were also able to further decrease the number of data nodes in order to optimize searching the data and indexing the results. For example, since most data nodes in the original performance benchmark had insignificant usage other than indexing results, we tested a cluster with only 24 c5.2xlarge data nodes along with 1x r5.16xlarge extension node, a further 33% reduction. We believe that with further experimentation and fine tuning, you can find an even more optimal mix, depending on other resource requirements for the cluster.
+
+### 2. Increase indexing throughput
+
+While we have not run this experiment, we believe you can further independently fine-tune the cluster by increasing sharding and achieve better throughput in indexing AD results.
+
+### 3. Increase the rate of generating historical results
```

Comment:
"Increase the rate of historical results generation"? "Increase the historical results generation rate"?

Reply:
I'm fine with the second version.